codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ConstraintsMemberShapeTest.kt

```diff
@@ -0,0 +1,1012 @@
+            "patternUnion",
+            "patternUnionOverride",
+            "patternMap",
+            "patternMapOverride",
+        ).forEach(::checkFieldTargetRemainsSame)
+
+        checkMemberShapeIsSame(
+            transformedModel,
+            outputModelOnly,
+            "constrainedMemberShape#StructWithConstrainedMember\$long",
+            "constrainedMemberShape#StructWithConstrainedMember\$long",
+        )
+    }
+
+    @Test
+    fun `constrained members should have a different target now`() {
+        val transformedModel = loadModel(outputModelOnly)
+        checkMemberShapeChanged(
+            transformedModel,
+            outputModelOnly,
+            "constrainedMemberShape#PatternStringListOverride\$member",
+            "constrainedMemberShape#PatternStringListOverride\$member",
+        )
+
+        fun checkSyntheticFieldTargetChanged(fieldName: String) {
+            checkMemberShapeChanged(
+                transformedModel,
+                outputModelOnly,
+                "constrainedMemberShape.synthetic#OperationUsingGetOutput\$$fieldName",
+                "constrainedMemberShape#OperationUsingGetOutput\$$fieldName",
+            ) {
+                "constrained member $fieldName should have been changed into a new type."
+            }
+        }
+
+        fun checkFieldTargetChanged(memberNameWithContainer: String) {
+            checkMemberShapeChanged(
+                transformedModel,
+                outputModelOnly,
+                "constrainedMemberShape#$memberNameWithContainer",
+                "constrainedMemberShape#$memberNameWithContainer",
+            ) {
+                "constrained member $memberNameWithContainer should have been changed into a new type."
+            }
+        }
+
+        setOf(
+            "constrainedLong",
+            "constrainedByte",
+            "constrainedShort",
+            "constrainedInteger",
+            "constrainedString",
+            "requiredConstrainedString",
+            "requiredConstrainedLong",
+            "requiredConstrainedByte",
+            "requiredConstrainedInteger",
+            "requiredConstrainedShort",
+            "constrainedPatternString",
+        ).forEach(::checkSyntheticFieldTargetChanged)
+
+        setOf(
+            "StructWithConstrainedMember\$lat",
+            "PatternMapOverride\$key",
+            "PatternMapOverride\$value",
+            "PatternStringListOverride\$member",
+        ).forEach(::checkFieldTargetChanged)
+    }
+
+    @Test
+    fun `extra trait on a constrained member should remain on it`() {
+        val transformedModel = loadModel(outputModelOnly)
+        checkShapeHasTrait(
+            transformedModel,
+            outputModelOnly,
+            "constrainedMemberShape.synthetic#OperationUsingGetOutput\$constrainedPatternString",
+            "constrainedMemberShape#OperationUsingGetOutput\$constrainedPatternString",
+            "aws.api#data",
```

Comment:
You can pass in `software.amazon.smithy.aws.traits.DataTrait` directly. You can make `checkShapeHasTrait` take in a `software.amazon.smithy.model.traits.Trait`, and use `memberShape.expectTrait<Trait>()` inside, instead of having to always manually construct the trait using `ShapeId.from(traitName)`.

Reply:
I think we should also test that `required` remains.